apps/studio/data/table-rows/table-rows-count-query.ts

```diff
@@ -9,31 +9,87 @@ import { formatFilterValue } from './utils'
 type GetTableRowsCountArgs = {
   table?: SupaTable
   filters?: Filter[]
+  enforceExactCount?: boolean
   impersonatedRole?: ImpersonationRole
 }
 
-export const getTableRowsCountSqlQuery = ({ table, filters = [] }: GetTableRowsCountArgs) => {
-  const query = new Query()
+export const THRESHOLD_COUNT = 50000
+const COUNT_ESTIMATE_SQL = `
+CREATE OR REPLACE FUNCTION pg_temp.count_estimate(
+    query text
+) RETURNS integer LANGUAGE plpgsql AS $$
+DECLARE
+    plan jsonb;
+BEGIN
+    EXECUTE 'EXPLAIN (FORMAT JSON)' || query INTO plan;
+    RETURN plan->0->'Plan'->'Plan Rows';
+END;
+$$;
+`.trim()
 
-  if (!table) {
-    return ``
-  }
+export const getTableRowsCountSqlQuery = ({
+  table,
+  filters = [],
+  enforceExactCount = false,
+}: GetTableRowsCountArgs) => {
+  if (!table) return ``
+
+  if (enforceExactCount) {
+    const query = new Query()
+    let queryChains = query.from(table.name, table.schema ?? undefined).count()
+    filters
+      .filter((x) => x.value && x.value != '')
+      .forEach((x) => {
+        const value = formatFilterValue(table, x)
+        queryChains = queryChains.filter(x.column, x.operator, value)
+      })
+    return `select (${queryChains.toSql().slice(0, -1)}), false as is_estimate;`
+  } else {
+    const selectQuery = new Query()
+    let selectQueryChains = selectQuery.from(table.name, table.schema ?? undefined).select('*')
+    filters
+      .filter((x) => x.value && x.value != '')
+      .forEach((x) => {
+        const value = formatFilterValue(table, x)
+        selectQueryChains = selectQueryChains.filter(x.column, x.operator, value)
+      })
+    const selectBaseSql = selectQueryChains.toSql()
 
-  let queryChains = query.from(table.name, table.schema ?? undefined).count()
-  filters
-    .filter((x) => x.value && x.value != '')
-    .forEach((x) => {
-      const value = formatFilterValue(table, x)
-      queryChains = queryChains.filter(x.column, x.operator, value)
-    })
+    const countQuery = new Query()
+    let countQueryChains = countQuery.from(table.name, table.schema ?? undefined).count()
+    filters
+      .filter((x) => x.value && x.value != '')
+      .forEach((x) => {
+        const value = formatFilterValue(table, x)
+        countQueryChains = countQueryChains.filter(x.column, x.operator, value)
+      })
+    const countBaseSql = countQueryChains.toSql().slice(0, -1)
 
-  const sql = queryChains.toSql()
+    const sql = `
+${COUNT_ESTIMATE_SQL}
 
-  return sql
+with approximation as (
+    select reltuples as estimate
+    from pg_class
+    where oid = ${table.id}
+)
+select 
+  case 
+    when estimate = -1 then (select pg_temp.count_estimate('${selectBaseSql}'))
+    when estimate > ${THRESHOLD_COUNT} then ${filters.length > 0 ? `pg_temp.count_estimate('${selectBaseSql}')` : 'estimate'}
```

Comment:
Single quotes in these two need to be escaped (for example `select * from posts where title = 'My Post'` should become `select * from posts where title = ''My Post''`

Reply:
ahh good catch! will fix these 🙏